### DISCUSSION TITLE
`jotai-molecule` and SSR

### DISCUSSION THREAD

**Discussion by sari-shipt** [pmndrs/jotai#2046] [Q&A]
Hello! I'm working on a feature that requires the usage of limited scoping using jotai molecules and ScopeProvider. However, my page is SSRed and using `useHydrateAtoms` on the atom out of the molecule basically removes the benefit of the `ScopeProvider`. Is there a known workaround for this?

**Comment by dai-shi** [comment]
@loganvolkers ?

**Comment by loganvolkers** [comment]
@sari-shipt can you describe a bit more about what you're trying to achieve? I don't have a known workaround, but I have a theory.

You would return your server-state alongside your atoms  in your molecule.

```ts
const ssrMolecule = molecule((getMol,getScope)=>{
   
   // A user ID scope
   const userId = getScope(userScope);

   // ⭐ Your SSR hydration? In theory you know how to look this up
   const serverCount = getCountFromServer(userId);

   // User-scoped atom
   const countAtom = atom(0);

   return {
      userId,
      countAtom,
      ssrAtoms: [[countAtom, serverCount]],
      // ... other things here
   }
});


function useUserSsrMolecule(){
   const {ssrAtoms} = useMolecule(ssrMolecule);
  
   // ⭐ makes hydration happen
   useHydrateAtoms(ssrAtoms);
   
   //... see https://jotai.org/docs/utilities/ssr
}
```

**Reply by SariSabouh** [reply]
Wouldn't this create a new molecule every time and also create a non-Scoped atom, @loganvolkers ?

The `useHydrateAtoms` will hydrate the `ssrAtoms` to the global Jotai cache and any file that calls that atom with a `useAtom` or `jotaiStore.get` would have access to that atom and its value, removing the benefit of using the `ScopeProvider scope={customScope} value={myLimitedValue}`.

I found a workaround by creating a component that is only purpose is to hydrate the atoms and set the scope before processing any children. For example, it previously looked something like this:

SSR
```typescript
const myAtom = 'value'
const mySoonToBeScopedAtom = 'another value'
```

Old Component Structure
`index.tsx`
```typescript
export const Component = (ssrAtoms, scopedAtom) => {
useHydrateAtoms(ssrAtoms)
// business logic that depends on the scope
return (
<ScopeProvider scope={myScope} value={scopedAtom}>
// rest of components
```

New Component Structure
`index.tsx`
```typescript
export const Component = (atoms, scopedAtom) => {
  useHydrateAtoms(atoms)
  return (
    <ScopeProvider scope={MyScope} value={scopedAtom}>
      <OldIndexFileThatDoesNotManageAtoms />
    </ScopeProvider>
  )
}
```

What are your thoughts on this approach, @dai-shi , @loganvolkers

**Reply by loganvolkers** [reply]
Can you please update your example to show what you're using scope for?

I think that's the missing piece to help me understand what you're trying to achieve.

Scoped molecules are a powerful way of creating and memoizing atoms. They can work with SSR and other store technologies besides jotai and react.

**Reply by SariSabouh** [reply]
I'm using as the docs suggest to set a limited scoped value based on a ScopeProvider:
```typescript
<ScopeProvider scope={InitialCountScope} value={1}>
      <Counter />
      <Counter />
    </ScopeProvider>
<ScopeProvider scope={InitialCountScope} value={2}>
      <OtherCounter />
      <OtherCounter />
    </ScopeProvider>
```

In that case as I understood, Counter value will be 1  and it will only be controlled within that scope. So if I change Counter through UI then OtherCounter will not be impacted, correct?

Side Q: No way to access newly set atoms within scope/molecule through the `jotaiStore`?

**Reply by loganvolkers** [reply]
@sari-shipt sorry I'm struggling to understand your problem with these abstract code samples.

Would it be possible for you to put together a complete codesandbox or other self contained example?

I suspect that you might need help with scopes more than SSR.

**Reply by SariSabouh** [reply]
Scopes are meant to provide a scoped value for an atom that all components under that `ScopeProvider`. So if there is a `<ScopeProvider scope={aScope} value={initialValue}`, then all children of `ScopeProvider` will have the value as `initialValue` for that scoped atom. But if that same atom is accessed outside of the `ScopeProvider`, then its value will be defaulted to whatever the `createScope` is set at. Is that correct?

**Comment by SariSabouh** [comment]
Any follow ups on this, @loganvolkers ?

Side note: @dai-shi - Have you thought about having a Context like solution applied through Jotai? I know that the docs have a rough image on how Jotai can replace Context, but it does not fully replace it due to being able to access atoms anywhere. I have a starter solution in my project that fully replaces Context and reduces boilerplate that also utilizes the `ScopeProvider`. So I'm curious if this is something already in the works :)

**Reply by dai-shi** [reply]
It replaces useState+useContext "global" use case, but doesn't fully replace contextual use case. That's why jotai-molecules was born.
We discussed something about a smaller version of jotai-molecules that only covers a smaller part of the use case. But, there has been no activities. If you are interested, you can share your idea.

**Reply by SariSabouh** [reply]
Where can I share the idea @dai-shi ?

**Reply by dai-shi** [reply]
Here in this thread. You can open a new discussion if you want.

**Reply by SariSabouh** [reply]
So I'm thinking something within these lines but better as it will be new code to the`jotai` library directly:

Introduce a new useScopedAtom that is only usable if a <ScopeProvider /> has been initialized with a default value. Something like this

useScopedAtom:
```typescript
export const useScopedAtom = <T extends AnyWritableAtom>(
  scopedMolecule: Molecule<T>
) => {
  const scopedAtom = useMolecule(scopedMolecule)
  const [scopedAtomValue, setScopedAtom] = useAtom<T>(scopedAtom)

  if (scopedAtomValue === DefaultErrorType) {
    const scopeError = `You cannot use ${scopedAtom.debugLabel} out of its scope`
    throw new Error(scopeError)
  }

  return [scopedAtomValue, setScopedAtom] as const
}
```

Limited Scope User
```typescript
<ScopeProvider scope={selectedPaymentScope} value={defaultFValue}>
  <ChildrenWithAccessToScopedAtom />
</ScopeProvider>
```

Atom Identification as of now. This is the part where I want to change things even more to make it more dynamic as this is purely built off the existing `jotai` and `jotai-molecule` libraries

```typescript
export const acSope = createScope<SelectedPaymentMethod[] | null>(
  null
)
export const atMolecule = molecule((_getMol, getScope) => {
  const thingScoped = getScope(selectedPaymentScope)
  const anAtom = atom(thingScoped)
  return anAtom
})
```

**Reply by dai-shi** [reply]
> I want to change things even more to make it more dynamic as this is purely built off the existing `jotai` and `jotai-molecule` libraries

I'm not sure if I fully understand your point, but how would it look like, and would you have an implementation idea?

**Reply by SariSabouh** [reply]
I do not have an exact solution to make it more dynamic, yet, as this is just a prototype I built for my project that applies the required Context-like limitations. But I would like to make it even lesser boilerplate for the end developer. Something that will require the developer to create one scoped atom that implies the scope, rather than having to create both the scope and the molecule to generate the atom, but I do not have that solution in place yet. 

What are your thoughts and comments on the solution I provided?

**Reply by dai-shi** [reply]
What I'm not sure is if you are suggesting something less than jotai-molecules or something more. So far, it sounds like jotai-molecules is just for you.

**Reply by SariSabouh** [reply]
It would be something less. jotai-molecules definitely did the needed job for my project but it required setup work within the project itself to reach a Context-like solution, like creating that useScopedAtom. With ReactContext you are only allowed to access the variables within the context provider, while jotai-molecules still allows global access if a ScopeProvider is not provided.

**Reply by dai-shi** [reply]
We are on the same page in that sense. 

Note that bare React Context still falls back to default value and it makes sense it works without providers. What else you differentiate from molecules?

**Reply by SariSabouh** [reply]
I think really that’s the main thing, jotai-molecule is great. Only thing I’m adding is just a more restricted, scoped atom solution, with lesser boilerplate.

**Reply by dai-shi** [reply]
Let's say we had an imaginary library `jotai-scope`. What would be the usage which is less boilerplate?

**Reply by SariSabouh** [reply]
As I do not have this coded or details finalized, but the usage would be something like this:

1. user creates a scoped atom. This scoped atom is only usable under a provider context. 
2. User creates a provider that expects a default value, and the scoped atom associated, similar to the ScopedProvider. 
3. User uses a new hook that will return the scoped atom, or throw if requested outside of scoped access.
4. Maybe add support to jotaiStore for access outside of books and easier unit tests.

**Reply by dai-shi** [reply]
Could you please create a small example with `jotai-molecules` and add comments where you feel boilerplate then? That would be helpful to get some ideas. You can fork https://csb.jotai.org.

**Reply by SariSabouh** [reply]
Sure. I’ll add it sometime this week 👍🏼

**Reply by SariSabouh** [reply]
Hi @dai-shi - Please checkout a quick demo and desired behavior within this sandbox: https://codesandbox.io/s/funny-brahmagupta-wcvgq2

It will show the desired usage with comments, and how similar this is to `jotai-molecules` with minor differences on scoping.

**Reply by dai-shi** [reply]
How about this? https://codesandbox.io/s/brave-http-myq73f?file=/src/App.tsx

**Reply by SariSabouh** [reply]
Ya I like that too! Only updates would be to change the atom creation to a scope specific typed, rather than your standard atom. In other words, instead of `atom(false)` to `scopedAtom<ExpectedType>()` and its default value OOTB would be an error type.

**Reply by SariSabouh** [reply]
If you agree with a solution like this, I would like to give it a shot myself

**Reply by dai-shi** [reply]
Please give it a try. Though, for such use cases, I don't think we need a library or a util. An advanced recipe might be a good fit.

**Reply by SariSabouh** [reply]
To be clear, you prefer to not have the solution we got to in the codesandbox with the `useScopedAtom` hook, for example, or a new `scoepdAtom` as a part of the library? And instead, just update the documentation and provide an advanced recipe?

**Reply by dai-shi** [reply]
I think `useScopedAtom` in the codesandbox is one user-land solution for the user requirement. It shouldn't be something in the core library.

For example, here's another solution which I may prefer: https://codesandbox.io/s/weathered-snow-5y2kh5?file=/src/App.tsx
(except for `any`, which I think should be resolved.)

**Reply by SariSabouh** [reply]
Got you. So any throwing or restriction applying should be done on the user side per component, same as you have with the Counter component. In that case, then ya this is not worth creating in the core library 👍 

Well thanks for your time reviewing the idea with me!
